src/parser/jobs/smn/modules/Ruin4.tsx

```diff
@@ -0,0 +1,137 @@
+
+const SEVERITY_STACK_COUNT = {
+	1: SEVERITY.MEDIUM,
+	5: SEVERITY.MAJOR,
+}
+
+const FURTHER_RUIN_PLAYER_ACTIONS = [
+	ACTIONS.ASSAULT_I_AERIAL_SLASH.id,
+	ACTIONS.ASSAULT_II_SLIIPSTREAM.id,
+	ACTIONS.ASSAULT_I_CRIMSON_CYCLONE.id,
+	ACTIONS.ASSAULT_II_FLAMING_CRUSH.id,
+	// NOTE: Titan Egi Assualt I Earthen Armor does not generate a Further Ruin stack
+	ACTIONS.ASSAULT_II_MOUNTAIN_BUSTER.id,
+]
+
+const FURTHER_RUIN_PET_ACTIONS = [
+	ACTIONS.AERIAL_SLASH.id,
+	ACTIONS.SLIPSTREAM.id,
+	ACTIONS.CRIMSON_CYCLONE.id,
+	ACTIONS.FLAMING_CRUSH.id,
+	// NOTE: Titan Egi Assualt I does not generate a Further Ruin stack
+	ACTIONS.MOUNTAIN_BUSTER.id,
+]
+
+const MAX_FURTHER_RUIN_COUNT = 4
+const END_OF_FIGHT_LEEWAY = 5000
+
+export default class Ruin4 extends Module {
+	static handle = 'ruin4'
+	static title = t('smn.ruin-iv.title')`Ruin IV`
+
+	@dependency private suggestions!: Suggestions
+	@dependency private invuln!: Invulnerability
+
+	private currentStackCount = 0
+	private earthenArmorCount = 0
+	private overage = 0
+	private playerSkillCount = 0
+	private petSkillCount = 0
+
+	protected init() {
+		this.addHook('cast', {by: 'player'}, this.onPlayerCast)
+		this.addHook('cast', {by: 'pet'}, this.onPetCast)
+		this.addHook('death', {to: 'player'}, this.onDeath)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onPlayerCast(event: CastEvent) {
+		if (event.ability.guid === ACTIONS.RUIN_IV.id) {
+			if (this.currentStackCount > 0) { this.currentStackCount-- }
+		} else if (FURTHER_RUIN_PLAYER_ACTIONS.includes(event.ability.guid)) {
+			// Do not flag for player skills used right at the end of the fight when the
+			// pet may not have time to use the skill.
+			const fightTimeRemaining = this.parser.fight.end_time - event.timestamp
+			if (fightTimeRemaining > END_OF_FIGHT_LEEWAY) {
+				this.playerSkillCount++
+			}
+		} else if (event.ability.guid === ACTIONS.ASSAULT_I_EARTHEN_ARMOR.id) {
+			this.earthenArmorCount++
+		}
+	}
+
+	private onPetCast(event: CastEvent) {
+		if (FURTHER_RUIN_PET_ACTIONS.includes(event.ability.guid) &&
+			!this.invuln.getInvulnerableUptime('all', event.timestamp)) {
+			this.petSkillCount++
+			if (this.currentStackCount >= MAX_FURTHER_RUIN_COUNT) {
+				this.overage++
+			} else {
+				this.currentStackCount++
+			}
+		}
+	}
+
+	private onDeath() {
+		this.currentStackCount = 0
+	}
+
+	private onComplete() {
+		if (this.overage > 0) {
+			this.suggestions.add(new TieredSuggestion({
+				icon: STATUSES.FURTHER_RUIN.icon,
+				content: <Trans id="smn.ruin-iv.overage.content">
+					You used Egi Assault skills in a way that overcapped your stacks of <ActionLink {...STATUSES.FURTHER_RUIN}/>.
+				</Trans>,
+				tiers: SEVERITY_STACK_COUNT,
+				value: this.overage,
+				why: <Trans id="smn.ruin-iv.overage.why">
+					{this.overage} Further Ruin <Plural value={this.overage} one="stack was" other="stacks were"/> lost.
+				</Trans>,
+			}))
+		}
+
+		let numberLost = this.earthenArmorCount
```

Comment:
I'm not _entirely_ sure how comfortable I am considering Earthen Armor as a total loss, but can leave it for now. Would be good to go over these sort of details with Nem once we're both back from holidays and such.